helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -513,13 +513,13 @@ private static PipelineRegistry createDefaultRegistry(String pipelineName) {
       rebalancePipeline.addStage(new BestPossibleStateCalcStage());
       // Need to add MaintenanceRecoveryStage here because MAX_PARTITIONS_PER_INSTANCE check could
       // only occur after IntermediateStateCalcStage calculation
-      rebalancePipeline.addStage(new MaintenanceRecoveryStage());
       rebalancePipeline.addStage(new MessageGenerationPhase());
       rebalancePipeline.addStage(new MessageSelectionStage());
       // The IntermediateStateCalcStage should be applied after message selection
       // Messages are throttled already removed by IntermediateStateCalcStage in MessageSelection output
       rebalancePipeline.addStage(new IntermediateStateCalcStage());
       rebalancePipeline.addStage(new MessageThrottleStage());
+      rebalancePipeline.addStage(new MaintenanceRecoveryStage());
```

Comment:
Because they are correlated. For phase2, we would like to combine IntermediateStage and MessageThrottling together. Thus the result can impact the decision of MaintenanceRecovery As well.

Reply:
Thanks for the clarification. Don't forget to move and update the comment